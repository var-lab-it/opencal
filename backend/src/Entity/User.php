<?php

declare(strict_types=1);

namespace App\Entity;

use ApiPlatform\Metadata\ApiResource;
use ApiPlatform\Metadata\Get;
use ApiPlatform\Metadata\Patch;
use App\Repository\UserRepository;
use App\State\CurrentUserProvider;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Safe\DateTimeImmutable;
use Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface;
use Symfony\Component\Security\Core\User\UserInterface;
use Symfony\Component\Serializer\Attribute\Groups;
use Symfony\Component\Validator\Constraints as Assert;

#[ApiResource(
    operations: [
        new Get(
            uriTemplate: '/me',
            security: "is_granted('IS_AUTHENTICATED_FULLY')",
            provider: CurrentUserProvider::class,
        ),
        new Patch(
            uriTemplate: '/me/{id}',
            security: "is_granted('IS_AUTHENTICATED_FULLY')",
        ),
    ],
    normalizationContext: [
        'groups' => [
            'me:read',
        ],
    ],
    denormalizationContext: [
        'groups' => [
            'me:write',
        ],
    ],
)]
#[ORM\Entity(repositoryClass: UserRepository::class)]
#[ORM\HasLifecycleCallbacks]
#[ORM\UniqueConstraint(name: 'UNIQ_IDENTIFIER_EMAIL', fields: ['email'])]
class User implements UserInterface, PasswordAuthenticatedUserInterface
{
    public const string ROLE_USER  = 'ROLE_USER';
    public const string ROLE_ADMIN = 'ROLE_ADMIN';

    #[Groups(['me:read', 'me:write'])]
    #[ORM\Column]
    #[ORM\GeneratedValue]
    #[ORM\Id]
    private ?int $id = null;

    #[Assert\Email]
    #[Assert\NotBlank]
    #[Groups(['me:read', 'me:write'])]
    #[ORM\Column(length: 180)]
    private string $email;

    /** @var list<string> The user roles */
    #[Groups(['me:read'])]
    #[ORM\Column]
    private array $roles = [];

    /** @var string The hashed password */
    #[ORM\Column]
    private string $password;

    #[Assert\NotBlank]
    #[Groups(['me:read', 'me:write', 'event_type:read'])]
    #[ORM\Column(length: 255)]
    private string $givenName;

    #[Assert\NotBlank]
    #[Groups(['me:read', 'me:write', 'event_type:read'])]
    #[ORM\Column(length: 255)]
    private string $familyName;

    /** @var Collection<int, EventType> */
    #[ORM\OneToMany(targetEntity: EventType::class, mappedBy: 'host')]
    private Collection $eventTypes;

    /** @var Collection<int, Unavailability> */
    #[ORM\OneToMany(targetEntity: Unavailability::class, mappedBy: 'user')]
    private Collection $unavailabilities;

    /** @var Collection<int, Availability> */
    #[ORM\OneToMany(targetEntity: Availability::class, mappedBy: 'user')]
    private Collection $availabilities;

    /** @var Collection<int, CalDavAuth> */
    #[ORM\OneToMany(targetEntity: CalDavAuth::class, mappedBy: 'user')]
    private Collection $calDavAuths;

    #[ORM\Column]
    private bool $enabled = false;

    #[ORM\Column]
    private \DateTimeImmutable $createdAt;

    #[ORM\Column]
    private \DateTimeImmutable $updatedAt;

    public function __construct()
    {
        $this->eventTypes       = new ArrayCollection();
        $this->unavailabilities = new ArrayCollection();
        $this->availabilities   = new ArrayCollection();
        $this->calDavAuths      = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getEmail(): string
    {
        return $this->email;
    }

    public function setEmail(string $email): static
    {
        $this->email = $email;

        return $this;
    }

    /**
     * A visual identifier that represents this user.
     *
     * @see UserInterface
     */
    public function getUserIdentifier(): string
    {
        /** @var non-empty-string $email */
        $email = $this->email;

        return $email;
    }

    /** @see UserInterface */
    public function getRoles(): array
    {
        $roles   = $this->roles;
        $roles[] = 'ROLE_USER';

        return \array_unique($roles);
    }

    /** @param list<string> $roles */
    public function setRoles(array $roles): static
    {
        $this->roles = $roles;

        return $this;
    }

    public function addRole(string $role): static
    {
        if (!\in_array($role, $this->roles, true)) {
            $this->roles[] = $role;
        }

        return $this;
    }

    /** @see PasswordAuthenticatedUserInterface */
    public function getPassword(): string
    {
        return $this->password;
    }

    public function setPassword(string $password): static
    {
        $this->password = $password;

        return $this;
    }

    /** @see UserInterface */
    public function eraseCredentials(): void
    {
    }

    public function getGivenName(): string
    {
        return $this->givenName;
    }

    public function setGivenName(string $givenName): static
    {
        $this->givenName = $givenName;

        return $this;
    }

    public function getFamilyName(): string
    {
        return $this->familyName;
    }

    public function setFamilyName(string $familyName): static
    {
        $this->familyName = $familyName;

        return $this;
    }

    /** @return Collection<int, EventType> */
    public function getEventTypes(): Collection
    {
        return $this->eventTypes;
    }

    public function addEventType(EventType $eventType): static
    {
        if (!$this->eventTypes->contains($eventType)) {
            $this->eventTypes->add($eventType);
            $eventType->setHost($this);
        }

        return $this;
    }

    public function removeEventType(EventType $eventType): static
    {
        if ($this->eventTypes->removeElement($eventType)) {
            // set the owning side to null (unless already changed)
            if ($eventType->getHost() === $this) {
                $eventType->setHost(null);
            }
        }

        return $this;
    }

    /** @return Collection<int, Unavailability> */
    public function getUnavailabilities(): Collection
    {
        return $this->unavailabilities;
    }

    public function addUnavailability(Unavailability $recurringUnavailability): static
    {
        if (!$this->unavailabilities->contains($recurringUnavailability)) {
            $this->unavailabilities->add($recurringUnavailability);
            $recurringUnavailability->setUser($this);
        }

        return $this;
    }

    public function removeUnavailability(Unavailability $recurringUnavailability): static
    {
        if ($this->unavailabilities->removeElement($recurringUnavailability)) {
            // set the owning side to null (unless already changed)
            if ($recurringUnavailability->getUser() === $this) {
                $recurringUnavailability->setUser(null);
            }
        }

        return $this;
    }

    /** @return Collection<int, Availability> */
    public function getAvailabilities(): Collection
    {
        return $this->availabilities;
    }

    public function addAvailability(Availability $availability): static
    {
        if (!$this->availabilities->contains($availability)) {
            $this->availabilities->add($availability);
            $availability->setUser($this);
        }

        return $this;
    }

    public function removeAvailability(Availability $availability): static
    {
        if ($this->availabilities->removeElement($availability)) {
            // set the owning side to null (unless already changed)
            if ($availability->getUser() === $this) {
                $availability->setUser(null);
            }
        }

        return $this;
    }

    /** @return Collection<int, CalDavAuth> */
    public function getCalDavAuths(): Collection
    {
        return $this->calDavAuths;
    }

    public function addCalDavAuth(CalDavAuth $calDavAuth): static
    {
        if (!$this->calDavAuths->contains($calDavAuth)) {
            $this->calDavAuths->add($calDavAuth);
            $calDavAuth->setUser($this);
        }

        return $this;
    }

    public function removeCalDavAuth(CalDavAuth $calDavAuth): static
    {
        if ($this->calDavAuths->removeElement($calDavAuth)) {
            // set the owning side to null (unless already changed)
            if ($calDavAuth->getUser() === $this) {
                $calDavAuth->setUser(null);
            }
        }

        return $this;
    }

    public function isEnabled(): bool
    {
        return $this->enabled;
    }

    public function setEnabled(bool $enabled): static
    {
        $this->enabled = $enabled;

        return $this;
    }

    public function getCreatedAt(): \DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function setCreatedAt(\DateTimeImmutable $createdAt): static
    {
        $this->createdAt = $createdAt;

        return $this;
    }

    public function getUpdatedAt(): \DateTimeImmutable
    {
        return $this->updatedAt;
    }

    public function setUpdatedAt(\DateTimeImmutable $updatedAt): static
    {
        $this->updatedAt = $updatedAt;

        return $this;
    }

    #[ORM\PrePersist]
    public function setCreatedAtValue(): void
    {
        $this->createdAt = new DateTimeImmutable();
    }

    #[ORM\PrePersist]
    #[ORM\PreUpdate]
    public function setUpdatedAtValue(): void
    {
        $this->updatedAt = new DateTimeImmutable();
    }
}
