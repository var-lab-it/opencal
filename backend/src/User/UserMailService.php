<?php

declare(strict_types=1);

namespace App\User;

use App\Entity\User;
use App\Mail\MailService;
use Symfony\Contracts\Translation\TranslatorInterface;

class UserMailService
{
    public function __construct(
        private readonly MailService $mailService,
        private readonly TranslatorInterface $translator,
        private readonly string $locale,
        private readonly string $frontendDomain,
        private readonly bool $useSSL,
    ) {
    }

    public function sendPasswordResetEmail(User $user): void
    {
        $params = [
            '{user_name}' => $user->getGivenName(),
            '{reset_url}' => \sprintf(
                '%s/password/reset/%s/%s',
                $this->getFrontendUrl(),
                $user->getPasswordResetToken(),
                $user->getEmail(),
            ),
        ];

        $this->mailService->sendEmail(
            $this->translator->trans('mails.password_request.subject', [], 'messages', $this->locale),
            $this->translator->trans('mails.password_request.message', $params, 'messages', $this->locale),
            $user->getEmail(),
            \sprintf(
                '%s %s',
                $user->getGivenName(),
                $user->getFamilyName(),
            ),
        );
    }

    protected function getFrontendUrl(): string
    {
        $protocol = $this->useSSL ? 'https' : 'http';

        return \sprintf(
            '%s://%s',
            $protocol,
            $this->frontendDomain,
        );
    }
}
