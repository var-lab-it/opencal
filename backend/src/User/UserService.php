<?php

declare(strict_types=1);

namespace App\User;

use App\Entity\User;
use App\Repository\UserRepository;
use Doctrine\ORM\EntityManagerInterface;

class UserService
{
    public function __construct(
        private readonly EntityManagerInterface $entityManager,
        private readonly UserRepository $userRepository,
    ) {
    }

    public function createUser(): User
    {
        $user = new User();
        $user
            ->setEnabled(true)
            ->setRoles([User::ROLE_USER]);

        return $user;
    }

    public function saveUser(User $user): void
    {
        $this->entityManager->persist($user);
        $this->entityManager->flush();
    }

    public function isEmailUsed(string $email): bool
    {
        $existingUser = $this->userRepository->findOneByEmail($email);

        return $existingUser instanceof User;
    }
}
