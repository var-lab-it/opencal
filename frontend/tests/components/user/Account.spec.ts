import {describe, expect, test, vi} from "vitest";
import {render} from "vitest-browser-vue";
import Account from "../../../src/components/user/Account.vue";
import {User} from "../../../src/types/User";

const userMock: User = {
    email: 'test@test.tld',
    id: 123,
    familyName: 'Hans',
    givenName: 'Doe',
    locale: 'en_GB',
    roles: [],
}

vi.mock("vue-i18n", () => {
    return {
        useI18n: () => ({
            t: (msg: string) =>
                ({}[msg] || msg),
        }),
    };
});

vi.mock("../../src/composables/CurrentUser.ts", () => ({
    getCurrentUser: vi.fn,
}));

describe('account', () => {
    test('renders the account', async () => {
        const result = render(Account, {
            props: {
                user: userMock,
            },
            global: {
                mocks: {
                    $t: (msg: string) => msg,
                    useI18n: () => ({
                        t: (key: string) => key, // gibt Schlüssel zurück
                    }),

                }
            }
        });

        expect(result).toMatchSnapshot();
    });
})
