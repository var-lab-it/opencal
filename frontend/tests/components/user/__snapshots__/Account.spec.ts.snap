// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account > renders the account 1`] = `
{
  "baseElement": <body>
    
  


    <div>
      
      <h2
        class="mb-3"
      >
        account.headline
      </h2>
      <div
        class="spinner-border text-dark"
        role="status"
      />
      <div
        aria-hidden="true"
        aria-labelledby="editModalLabel"
        class="modal fade"
        id="editModal"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <h5
                class="modal-title"
                id="editModalLabel"
              >
                account.form.title
              </h5>
              <button
                class="btn-close"
                type="button"
              />
            </div>
            <div
              class="modal-body"
            >
              <form>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                  >
                    account.form.given_name
                  </label>
                  <input
                    class="form-control given-name"
                    type="text"
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                  >
                    account.form.family_name
                  </label>
                  <input
                    class="form-control family-name"
                    type="text"
                  />
                </div>
                <div
                  class="mb-3 border-bottom pb-3"
                >
                  <label
                    class="form-label"
                  >
                    account.form.email
                  </label>
                  <input
                    class="form-control email"
                    type="email"
                  />
                </div>
                <button
                  class="btn btn-primary"
                  id="submit-btn"
                  type="submit"
                >
                  account.form.buttons.save
                </button>
                <button
                  class="btn btn-outline-danger float-end"
                  type="button"
                >
                  account.form.buttons.cancel
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
      
    </div>
  </body>,
  "container": <div>
    
    <h2
      class="mb-3"
    >
      account.headline
    </h2>
    <div
      class="spinner-border text-dark"
      role="status"
    />
    <div
      aria-hidden="true"
      aria-labelledby="editModalLabel"
      class="modal fade"
      id="editModal"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
              id="editModalLabel"
            >
              account.form.title
            </h5>
            <button
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <form>
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                >
                  account.form.given_name
                </label>
                <input
                  class="form-control given-name"
                  type="text"
                />
              </div>
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                >
                  account.form.family_name
                </label>
                <input
                  class="form-control family-name"
                  type="text"
                />
              </div>
              <div
                class="mb-3 border-bottom pb-3"
              >
                <label
                  class="form-label"
                >
                  account.form.email
                </label>
                <input
                  class="form-control email"
                  type="email"
                />
              </div>
              <button
                class="btn btn-primary"
                id="submit-btn"
                type="submit"
              >
                account.form.buttons.save
              </button>
              <button
                class="btn btn-outline-danger float-end"
                type="button"
              >
                account.form.buttons.cancel
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
    
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholder": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`header > renders the header 1`] = `
{
  "baseElement": <body>
    
  


    <div>
      <nav
        class="navbar navbar-expand-lg bg-light sticky-top shadow-sm border-bottom"
      >
        <div
          class="container-fluid"
        >
          <router-link
            class="navbar-brand"
            to="/"
          >
            <!---->
             OpenCal 
          </router-link>
          <button
            aria-controls="navbarSupportedContent"
            aria-expanded="false"
            aria-label="Toggle navigation"
            class="navbar-toggler"
            data-bs-target="#navbarSupportedContent"
            data-bs-toggle="collapse"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="collapse navbar-collapse"
            id="navbarSupportedContent"
          >
            <ul
              class="navbar-nav ms-auto mb-2 mb-lg-0"
            >
              <li
                class="nav-item"
              >
                <router-link
                  active-class="active"
                  class="nav-link"
                  to="/"
                >
                  header.nav.dashboard
                </router-link>
              </li>
              <li
                class="nav-item"
              >
                <router-link
                  active-class="active"
                  class="nav-link"
                  to="/appointments"
                >
                  header.nav.appointments
                </router-link>
              </li>
              <li
                class="nav-item dropdown"
              >
                <a
                  aria-expanded="false"
                  class="nav-link dropdown-toggle"
                  data-bs-toggle="dropdown"
                  href="#"
                  role="button"
                >
                  <!---->
                   test@test.tld
                </a>
                <ul
                  class="dropdown-menu dropdown-menu-lg-end"
                >
                  <li>
                    <router-link
                      class="nav-link"
                      id="my-account-link"
                      to="/account"
                    >
                      <!---->
                       header.nav.account
                    </router-link>
                  </li>
                  <li
                    class="dropdown-divider"
                  />
                  <li>
                    <a
                      class="nav-link"
                      href="#"
                      id="logout-link"
                    >
                      <!---->
                       header.nav.logout
                    </a>
                  </li>
                </ul>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-lg bg-light sticky-top shadow-sm border-bottom"
    >
      <div
        class="container-fluid"
      >
        <router-link
          class="navbar-brand"
          to="/"
        >
          <!---->
           OpenCal 
        </router-link>
        <button
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-bs-target="#navbarSupportedContent"
          data-bs-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarSupportedContent"
        >
          <ul
            class="navbar-nav ms-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <router-link
                active-class="active"
                class="nav-link"
                to="/"
              >
                header.nav.dashboard
              </router-link>
            </li>
            <li
              class="nav-item"
            >
              <router-link
                active-class="active"
                class="nav-link"
                to="/appointments"
              >
                header.nav.appointments
              </router-link>
            </li>
            <li
              class="nav-item dropdown"
            >
              <a
                aria-expanded="false"
                class="nav-link dropdown-toggle"
                data-bs-toggle="dropdown"
                href="#"
                role="button"
              >
                <!---->
                 test@test.tld
              </a>
              <ul
                class="dropdown-menu dropdown-menu-lg-end"
              >
                <li>
                  <router-link
                    class="nav-link"
                    id="my-account-link"
                    to="/account"
                  >
                    <!---->
                     header.nav.account
                  </router-link>
                </li>
                <li
                  class="dropdown-divider"
                />
                <li>
                  <a
                    class="nav-link"
                    href="#"
                    id="logout-link"
                  >
                    <!---->
                     header.nav.logout
                  </a>
                </li>
              </ul>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholder": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
